recirq/otoc/utils.py

```diff
@@ -0,0 +1,253 @@
+    u33 = np.exp(1j * (delta_plus - phi))
+    return np.array([[1, 0, 0, 0], [0, u11, u12, 0], [0, u21, u22, 0],
+                     [0, 0, 0, u33]], dtype=complex)
+
+
+def fsim_to_angles(u_fsim: np.ndarray) -> Dict[str, float]:
+    """Converts a 2x2 FSIM unitary to five phases.
+
+    Args:
+        u_fsim: A 2x2 np.array corresponding to an FSIM unitary
+
+    Returns:
+        A dictionary with the keys being the name of a unitary phase and
+        value being the value of the phase.
+    """
+    u_fsim = u_fsim * np.exp(-1j * np.angle(u_fsim[0, 0]))
+    theta = np.arctan(np.abs(u_fsim[1, 2] / u_fsim[1, 1]))
+    delta_plus = np.angle(-u_fsim[1, 2] * u_fsim[2, 1])
+    phi = -np.angle(u_fsim[3, 3]) + delta_plus
+    delta_minus_off_diag = np.angle(
+        u_fsim[2, 1] / (-1j) / np.exp(1j * delta_plus / 2.0)) * 2.0
+    delta_minus_diag = np.angle(
+        u_fsim[1, 1] / np.exp(1j * delta_plus / 2.0)) * 2.0
+
+    angles = {'theta': theta,
+              'delta_plus': delta_plus,
+              'delta_minus_off_diag': delta_minus_off_diag,
+              'delta_minus_diag': delta_minus_diag,
+              'phi': phi}  # type: Dict[str, float]
+
+    return angles
+
+
+def generic_fsim_gate(fsim_angles: Dict[str, float],
+                      qubits: Tuple[cirq.GridQubit, cirq.GridQubit]
+                      ) -> List[cirq.OP_TREE]:
+    """Converts five FSIM phases to a list of Cirq ops.
+
+    Args:
+        fsim_angles:A dictionary with the keys being the name of a unitary
+            phase and value being the value of the phase.
+        qubits: The pair of cirq.GridQubits to which the gates are applied.
+
+    Returns:
+        A list of gates (equivalent to an FSIM gate) acting on the two qubits.
+    """
+    q_0, q_1 = qubits
+    g_f = [cirq.Z(q_0) ** (-(fsim_angles['delta_minus_off_diag'] +
+                             fsim_angles['delta_minus_diag'] -
+                             2 * fsim_angles['delta_plus']) / np.pi / 4.0),
+           cirq.Z(q_1) ** ((fsim_angles['delta_minus_off_diag'] +
+                            fsim_angles['delta_minus_diag'] +
+                            2 * fsim_angles['delta_plus']) / np.pi / 4.0)
+           ]  # type: List[cirq.OP_TREE]
+
+    if not np.isclose(fsim_angles['phi'], 0):
+        g_f.append(cirq.CZ(q_0, q_1) ** (-fsim_angles['phi'] / np.pi))
+
+    if not np.isclose(fsim_angles['theta'], 0):
+        g_f.append(cirq.ISWAP(q_0, q_1) ** (
+                -fsim_angles['theta'] / (np.pi / 2.0)))
+
+    g_f.append(cirq.Z(q_0) ** (-(fsim_angles['delta_minus_diag'] -
+                                 fsim_angles['delta_minus_off_diag']) /
+                               np.pi / 4.0))
+    g_f.append(cirq.Z(q_1) ** ((fsim_angles['delta_minus_diag'] -
+                                fsim_angles['delta_minus_off_diag']) /
+                               np.pi / 4.0))
+    return g_f
+
+
+def cz_to_sqrt_iswap(qubit_0: cirq.GridQubit,
```

Comment:
Is this in Cirq or `cirq.google`? It probably should be

Reply:
Not sure where it is. I simplified this function and added the corrections in the XEB function itself.